src/block_storage/protocol.rs

```diff
@@ -0,0 +1,201 @@
+// Copyright 2024 Sandro-Alessio Gierens <sandro@gierens.de>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! JSON structures and protocol bits for the Block Storage API.
+
+#![allow(non_snake_case)]
+#![allow(missing_docs)]
+
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+protocol_enum! {
+    #[doc = "Possible volume statuses."]
+    enum VolumeStatus {
+        Creating = "creating",
+        Available = "available",
+        Reserved = "reserved",
+        Attaching = "attaching",
+        Detaching = "detaching",
+        InUse = "in-use",
+        Maintenance = "maintenance",
+        Deleting = "deleting",
+        AwaitingTransfer = "awaiting-transfer",
+        Error = "error",
+        ErrorDeleting = "error_deleting",
+        BackingUp = "backing-up",
+        RestoringBackup = "restoring-backup",
+        ErrorBackingUp = "error_backing-up",
+        ErrorRestoring = "error_restoring",
+        ErrorExtending = "error_extending",
+        Downloading = "downloading",
+        Uploading = "uploading",
+        Retyping = "retyping",
+        Extending = "extending"
+    }
+}
+
+protocol_enum! {
+    #[doc = "Available sort keys."]
+    enum VolumeSortKey {
+        CreatedAt = "created_at",
+        Id = "id",
+        Name = "name",
+        UpdatedAt = "updated_at"
+    }
+}
+
+impl Default for VolumeSortKey {
+    fn default() -> VolumeSortKey {
+        VolumeSortKey::CreatedAt
+    }
+}
+
+/// A volume attachment.
+#[derive(Debug, Clone, Deserialize)]
+pub struct VolumeAttachment {
```

Comment:
This is a part of the public API, does it make sense to make it as non-exhaustive to make sure we can add more fields without breaking anyone?

Reply:
What exactly do you mean by "non-exhaustive"?

I just added the fields that the cinder documentation show in an example, so would hope this is everything.

While checking the docs again I just realized, though, that the `host_name` should be an `Option`. Gonna fix that real quick.

Reply:
Ah now I know what you mean by non-exhaustive, saw it in the `protocol_enum` macro ... some of the Rust intricacies I still have to get familiar with :D ... I added `#[non_exhaustive]` to the struct.